src/_imagingft.c

```diff
@@ -748,73 +797,80 @@ font_render(FontObject* self, PyObject* args)
             x1 = im->xsize - xx;
 
-        if (mask) {
-            /* use monochrome mask (on palette images, etc) */
-            for (y = 0; y < glyph->bitmap.rows; y++) {
-                int yy = y + im->ysize - (PIXEL(glyph->metrics.horiBearingY) + ascender);
+        source = (unsigned char*) glyph->bitmap.buffer;
+        for (bitmap_y = 0; bitmap_y < glyph->bitmap.rows; bitmap_y++) {
+            if (horizontal_dir) {
+                yy = bitmap_y + im->ysize - (PIXEL(glyph->metrics.horiBearingY) + ascender);
                 yy -= PIXEL(glyph_info[i].y_offset);
-                if (yy >= 0 && yy < im->ysize) {
-                    /* blend this glyph into the buffer */
-                    unsigned char *target = im->image8[yy] + xx;
-                    int i, j, m = 128;
-                    for (i = j = 0; j < x1; j++) {
-                        if (j >= x0 && (source[i] & m))
+            } else {
+                yy = bitmap_y + PIXEL(y) + PIXEL(glyph->metrics.vertBearingY) + ascender;
+                yy += PIXEL(glyph_info[i].y_offset);
+            }
+            if (yy >= 0 && yy < im->ysize) {
+                // blend this glyph into the buffer
+                unsigned char *target = im->image8[yy] + xx;
+                if (mask) {
+                    // use monochrome mask (on palette images, etc)
+                    int j, k, m = 128;
+                    for (j = k = 0; j < x1; j++) {
+                        if (j >= x0 && (source[k] & m))
                             target[j] = 255;
                         if (!(m >>= 1)) {
                             m = 128;
-                            i++;
+                            k++;
                         }
                     }
-                }
-                source += glyph->bitmap.pitch;
-            }
-        } else {
-            /* use antialiased rendering */
-            for (y = 0; y < glyph->bitmap.rows; y++) {
-                int yy = y + im->ysize - (PIXEL(glyph->metrics.horiBearingY) + ascender);
-                yy -= PIXEL(glyph_info[i].y_offset);
-                if (yy >= 0 && yy < im->ysize) {
-                    /* blend this glyph into the buffer */
-
-                    int i;
-                    unsigned char *target = im->image8[yy] + xx;
-                    for (i = x0; i < x1; i++) {
-                        if (target[i] < source[i])
-                            target[i] = source[i];
+                } else {
+                    // use antialiased rendering
+                    int k;
+                    for (k = x0; k < x1; k++) {
+                        if (target[k] < source[k])
+                            target[k] = source[k];
                     }
                 }
-                source += glyph->bitmap.pitch;
             }
+            source += glyph->bitmap.pitch;
+        }
+        if (horizontal_dir) {
+            x += glyph_info[i].x_advance;
+        } else {
+            y -= glyph_info[i].y_advance;
```

Comment:
I think the condition is not needed, you can assign x and y unconditionally. In horizontal direction `y_advance` will be zero and in vertical direction `x_advance` will be zero.

Reply:
Okay. I've updated the commit.